codec-http/src/main/java/io/netty/handler/codec/http/HttpServerKeepAliveHandler.java

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright 2016 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.codec.http;
+
+import io.netty.channel.ChannelDuplexHandler;
+import io.netty.channel.ChannelFutureListener;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelPipeline;
+import io.netty.channel.ChannelPromise;
+
+import static io.netty.handler.codec.http.HttpUtil.*;
+
+/**
+ * HttpServerKeepAliveHandler helps close persistent connections when appropriate.
+ * <p>
+ * The server channel is expected to set the proper 'Connection' header if it can handle persistent connections. {@link
+ * HttpServerKeepAliveHandler} will automatically close the channel for any LastHttpContent that corresponds to a client
+ * request for closing the connection, or if the HttpResponse associated with that LastHttpContent requested closing the
+ * connection or didn't have a self defined message length.
+ * <p>
+ * Since {@link HttpServerKeepAliveHandler} expects {@link HttpObject}s it should be added after {@link HttpServerCodec}
+ * but before any other handlers that might send a {@link HttpResponse}. <blockquote>
+ * <pre>
+ *  {@link ChannelPipeline} p = ...;
+ *  ...
+ *  p.addLast("serverCodec", new {@link HttpServerCodec}());
+ *  p.addLast("httpKeepAlive", <b>new {@link HttpServerKeepAliveHandler}()</b>);
+ *  p.addLast("aggregator", new {@link HttpObjectAggregator}(1048576));
+ *  ...
+ *  p.addLast("handler", new HttpRequestHandler());
+ *  </pre>
+ * </blockquote>
+ */
+public class HttpServerKeepAliveHandler extends ChannelDuplexHandler {
+    private static final String MULTIPART_PREFIX = "multipart";
+
+    private boolean persistentConnection = true;
+    // Track pending responses to support client pipelining: https://tools.ietf.org/html/rfc7230#section-6.3.2
+    private int pendingResponses;
+
+    @Override
+    public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+        // read message and track if it was keepAlive
+        if (msg instanceof HttpRequest) {
+            final HttpRequest request = (HttpRequest) msg;
+            if (persistentConnection) {
+                pendingResponses += 1;
```

Comment:
I used to unconditionally increment `pendingResponses` but it made test case 7 fail once I started using it in the pipeline test.  If I make them consistent it will cause either test 7 to fail or 0-7 (except 2 and 4).  

The idea is once we know that this is no longer a persistent connection (for example the client sent `Connection: close`) then we should stop counting requests that are coming in.  It shouldn't really happen, but maybe a client sends a request to close the connection and then accidentally sends another request, in that case we will close the connection anyway as requested.